server/edd/load/migrations/0003_defaultunit.py

```diff
@@ -18,236 +20,210 @@ def bootstrap(apps, schema_editor):
-        type_name="Volume",
-        uuid="d8e9f1ea-9991-44c3-ba8e-5ee25f44aba7",
-    )
-    MeasurementType.objects.get_or_create(
-        type_group=GENERIC,
-        type_name="OUR",
-        uuid="3a7482bf-c259-48db-8534-31a336681f5e",
-    )
-    MeasurementType.objects.get_or_create(
-        type_group=GENERIC,
-        type_name="CER",
-        uuid="b55b5d30-cfab-450c-ab2e-fdee4bfbd602",
-    )
-    MeasurementType.objects.get_or_create(
-        type_group=GENERIC,
-        type_name="Volume",
-        uuid="d8e9f1ea-9991-44c3-ba8e-5ee25f44aba7",
-    )
-    MeasurementType.objects.get_or_create(
-        type_group=GENERIC, type_name="RQ", uuid="97ed0ed7-1eaa-4427-9c44-9712d5070e98",
-    )
-    MeasurementType.objects.get_or_create(
-        type_group=GENERIC,
-        type_name="Feed volume pumped",
-        uuid="bd88d980-c418-44be-b12d-4b1a6cda08d3",
-    )
-    MeasurementType.objects.get_or_create(
-        type_group=GENERIC,
-        type_name="Base volume pumped",
-        uuid="83c50be5-199f-4ef8-93fc-5db77b09e3c3",
-    )
-    MeasurementType.objects.get_or_create(
-        type_group=GENERIC,
-        type_name="Acid volume pumped",
-        uuid="15192dca-4863-475f-bf79-910bd722997a",
-    )
-    MeasurementType.objects.get_or_create(
-        type_group=GENERIC,
-        type_name="Antifoam volume pumped",
-        uuid="3a7ae3c2-c339-4c2c-821b-12d63dfcd39d",
-    )
-    MeasurementType.objects.get_or_create(
-        type_group=GENERIC,
-        type_name="Volume sampled",
-        uuid="0c88a7b1-ef9e-4271-9f54-e6e63869433c",
-    )
-    MeasurementType.objects.get_or_create(
-        type_group=GENERIC,
-        type_name="Dissolved Oxygen",
-        uuid="b65fb9bc-4e63-4617-b143-bf08f8b44699",
-    )
-
-
-def boostrap_measurement_unit(apps):
     MeasurementUnit = apps.get_model("main", "MeasurementUnit")
+    DefaultUnit = apps.get_model("load", "DefaultUnit")
     # constants on main.models.MeasurementType.Group
     GENERIC = "_"
-    MeasurementUnit.objects.get_or_create(type_group=GENERIC, unit_name="lpm")
-    MeasurementUnit.objects.get_or_create(type_group=GENERIC, unit_name="°C")
-    MeasurementUnit.objects.get_or_create(type_group=GENERIC, unit_name="rpm")
-    MeasurementUnit.objects.get_or_create(type_group=GENERIC, unit_name="°C")
-    MeasurementUnit.objects.get_or_create(type_group=GENERIC, unit_name="mL")
-    MeasurementUnit.objects.get_or_create(type_group=GENERIC, unit_name="mM/L/h")
-    MeasurementUnit.objects.get_or_create(
+    METABOLITE = "m"
+
+    C_unit, created = MeasurementUnit.objects.get_or_create(
+        type_group=GENERIC, unit_name="°C"
+    )
+    try:
+        temp_obj = MeasurementType.objects.get(
+            type_group=GENERIC, type_name="Temperature",
+        )
+    except MeasurementType.DoesNotExist:
+        uuid = uuid4()
+        temp_obj = MeasurementType(
+            type_group=GENERIC, type_name="Temperature", uuid=uuid,
+        )
+        temp_obj.save()
```

Comment:
This exact idiom is why `get_or_create` exists. [A reminder from the documentation](https://docs.djangoproject.com/en/3.2/ref/models/querysets/#get-or-create). In addition, the `save` method on `MeasurementType` already detects if a UUID is missing, and assigns one. So, these `try`-`except` blocks are unnecessary.

These lines would instead be:

```python
temperature, _created = MeasurementType.objects.get_or_create(
    type_group=GENERIC, type_name="Temperature",
)
```

I don't see any cases in this file where anything other than UUID is used, so each should get fixed as above, and then remove the `import uuid` at the top of the file, as nothing else should be using it at that point. This is a good test for ensuring you still have `pre-commit` properly set up on your clone, as the lint tools will complain about the unused import.

Note: the underscore in `_created` is mildly important, because stricter linters will generate warnings about an unused variable; adding the underscore in front is the conventional signal used, when unpacking a tuple return value, that the variable name is not going to be used later.

Also note: using `temp_obj` as a name is very misleading, as most will read it as "temporary object". My replacement example uses `temperature` to avoid this confusion.

Reply:
I just resolved this.

Reply:
I am getting an error after I made the changes and running the tests as below:

self = <django.db.backends.utils.CursorWrapper object at 0x7f95a01587c0>
sql = 'INSERT INTO "measurement_type" ("type_name", "short_name", "type_group", "provisional", "uuid", "alt_names", "type_source_id") VALUES (%s, %s, %s, %s, %s, %s::text[], %s) RETURNING "measurement_type"."id"'
params = ('Temperature', None, '_', False, None, [], ...)
ignored_wrapper_args = (False, {'connection': <django.db.backends.postgresql.base.DatabaseWrapper object at 0x7f95b59fb790>, 'cursor': <django.db.backends.utils.CursorWrapper object at 0x7f95a01587c0>})

    def _execute(self, sql, params, *ignored_wrapper_args):
        self.db.validate_no_broken_transaction()
        with self.db.wrap_database_errors:
            if params is None:
                # params default might be backend specific.
                return self.cursor.execute(sql)
            else:
>               return self.cursor.execute(sql, params)
E               django.db.utils.IntegrityError: null value in column "uuid" violates not-null constraint
E               DETAIL:  Failing row contains (2, Temperature, null, _, f, null, {}, null).

/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py:84: IntegrityError
========================================================================================================= short test summary info ==========================================================================================================
ERROR edd/load/tests/test_ambr_parser.py::test_AmbrExcelParser_success - django.db.utils.IntegrityError: null value in column "uuid" violates not-null constraint

It seems it is unable to create the uuid if not supplied while running tests and creating the test database.

Reply:
This works when I use the try except method.

Reply:
I just made the changes and squashed all the migration related commits, sending a new pull request to you.

Reply:
or maybe we can use this same pull request